### PR TITLE
Add a flat address space to `librustc_codegen_llvm`

### PR DESCRIPTION
**Pull request by YakoYakoYokuYoku**
To achieve compatibility with LLVM target machines which place allocas in another address space we need to the presence of a flat address space (an address space that is shared with other address spaces).

Also this PR updates #51576.

### CODE REVIEW

**Comment by cramertj** [comment]
r? @eddyb 
cc @alexcrichton

**Comment by mati865** [comment]
Changes to `llvm-project` and `cargo` submodules seem to be messed up.

**Comment by eddyb** [comment]
There's a merge commit that shouldn't exist - you might need to rebase.

**Comment by eddyb** [comment]
This seems to contain #51576, which was never shrunk enough for inclusion.
I'd suggest implementing it from scratch, casting everything that might not be in the flat address space to it right away, to avoid ever having to represent non-flat address spaces.

**Comment by YakoYakoYokuYoku** [comment]
Last night I was panicking that everything wasn't working because I've crashed miserably with the `nightly` release, now It's solved :cowboy_hat_face: .

While reviewing the code locally I've seen a lot of the code problems pointed out by @eddyb , like a few lines we could change:
```rust
let start = dest.project_index(&mut self, zero).llval;
let start = self.flat_addr_cast(start);
```
to something more like:
```rust
let start = self.flat_addr_cast(dest.project_index(&mut self, zero).llval);
```

I've left them untouched 'cause I wanted to see if the PR itself could work with `nightly` (and it worked with the `Linux mingw-check` in the first checks). As for now I'll start fixing the casting problems :wrench: .

**Comment by Alexendoo** [comment]
Ping from triage, any updates? @eddyb

**Comment by eddyb** [comment]
@Alexendoo Sorry, this was incorrectly tagged.
I don't think this is done, and there are merge commits that should be rebased out.

**Comment by Dylan-DPC** [comment]
@eddyb they have pushed a new commit. You can review it now.

**Comment by YakoYakoYokuYoku** [comment]
@eddyb the conflicts are now solved, you can review this freely :eyeglasses: :pencil2:

**Comment by edmilsonefs** [comment]
Hey! This is a ping from triage, we would like to know if you @eddyb could give us a few more minutes to update us from last changes made by @YakoYakoYokuYoku .

Thanks.

File: src/librustc_codegen_ssa/mir/place.rs [hunk 1]
```diff
@@ -67,6 +67,18 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
...
+    pub fn alloca_addr_space<Bx: BuilderMethods<'a, 'tcx, Value = V>>(
+        bx: &mut Bx,
+        layout: TyLayout<'tcx>,
+        name: &str
+    ) -> Self {
+        debug!("alloca({:?}: {:?})", name, layout);
+        assert!(!layout.is_unsized(), "tried to statically allocate unsized place");
+        let tmp = bx.alloca(bx.cx().backend_type(layout), name, layout.align.abi);
+        Self::new_sized(tmp, layout, layout.align.abi)
+    }
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/place.rs]: This is identical to `alloca` above though?

File: src/librustc_codegen_ssa/mir/block.rs [hunk 2]
```diff
@@ -292,7 +293,8 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 bx.load(addr, self.fn_ty.ret.layout.align.abi)
             }
         };
-        bx.ret(llval);
+        let llvalc = bx.flat_addr_cast(llval);
```
Comment (eddyb) [inline src/librustc_codegen_ssa/mir/block.rs]: IMO the only way this PR will be accepted without an RFC is if *everything* is cast to the flat address space *as soon as possible* (as opposed to when needed).

File: src/librustc_codegen_ssa/traits/type_.rs [hunk 3]
```diff
@@ -89,6 +125,48 @@ pub trait DerivedTypeMethods<'tcx>: BaseTypeMethods<'tcx> + MiscMethods<'tcx> {
...
+    }
+    fn type_ptr_to_alloca(&self, ty: Self::Type) -> Self::Type {
+        self.type_as_ptr_to(ty, self.alloca_addr_space())
+    }
+    fn type_ptr_to_const(&self, ty: Self::Type) -> Self::Type {
+        self.type_as_ptr_to(ty, self.const_addr_space())
+    }
+    fn type_ptr_to_mut(&self, ty: Self::Type) -> Self::Type {
+        self.type_as_ptr_to(ty, self.mutable_addr_space())
+    }
```
Comment (eddyb) [inline src/librustc_codegen_ssa/traits/type_.rs]: These methods shouldn't be added if they're unused.

File: src/librustc_codegen_llvm/common.rs [hunk 4]
```diff
@@ -356,6 +369,19 @@ pub fn val_ty(v: &'ll Value) -> &'ll Type {
...
+    let ty = val_ty(v);
+    if ty.kind() == TypeKind::Pointer {
+        Some(ty.address_space())
+    } else {
+        None
+    }
+}
+
+pub fn val_addr_space(v: &'ll Value) -> AddrSpaceIdx {
+    val_addr_space_opt(v).unwrap_or_default()
```
Comment (eddyb) [inline src/librustc_codegen_llvm/common.rs]: Reflecting LLVM values should be avoided as much as possible. New uses of `val_ty` shouldn't be added.

**Comment by ProgrammaticNajel** [comment]
Ping from triage. @YakoYakoYokuYoku any updates on this? Thanks

@rustbot modify labels to +S-waiting-on-author, -S-waiting-on-review

**Comment by YakoYakoYokuYoku** [comment]
Think It's better to prepare a proper RFC, closing this.

**Pull request closed without merging**